domain/user/state/state_test.go

```diff
@@ -11,74 +11,639 @@ import (
+	c.Assert(err, jc.ErrorIsNil)
+	jujuSystemCreatorUuid := jujuSystemUuid
+	jujuSystemUser := user.User{
+		Name:        "juju-system",
+		DisplayName: "juju-system",
+	}
+	err = st.AddUser(context.Background(), jujuSystemUuid, jujuSystemUser, jujuSystemCreatorUuid)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Add admin user.
+	userUUID, err := user.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	userCreatorUuid := jujuSystemUuid
+	adminUser := user.User{
+		Name:        "admin",
+		DisplayName: "admin",
+	}
+	err = st.AddUser(context.Background(), userUUID, adminUser, userCreatorUuid)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Add userToRemove.
+	userToRemoveUUID, err := user.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	userToRemove := user.User{
+		Name:        "userToRemove",
+		DisplayName: "userToRemove",
+	}
+	err = st.AddUser(context.Background(), userToRemoveUUID, userToRemove, userUUID)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Remove userToRemove.
+	err = st.RemoveUser(context.Background(), userToRemoveUUID)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Check that the user was removed correctly.
+	db := s.DB()
+
+	row := db.QueryRow(`
+SELECT removed
+FROM user
+WHERE uuid = ?
+	`, userToRemoveUUID)
+	c.Assert(row.Err(), jc.ErrorIsNil)
+
+	var removed bool
+	err = row.Scan(&removed)
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Check(removed, gc.Equals, true)
+}
+
+func (s *stateSuite) TestSetPasswordHash(c *gc.C) {
+	st := NewState(s.TxnRunnerFactory())
+
+	// Add juju-system user.
+	jujuSystemUuid, err := user.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	jujuSystemCreatorUuid := jujuSystemUuid
+	jujuSystemUser := user.User{
+		Name:        "juju-system",
+		DisplayName: "juju-system",
+	}
+	err = st.AddUser(context.Background(), jujuSystemUuid, jujuSystemUser, jujuSystemCreatorUuid)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Add admin user.
+	userUUID, err := user.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	userCreatorUuid := jujuSystemUuid
+	adminUser := user.User{
+		Name:        "admin",
+		DisplayName: "admin",
+	}
+	err = st.AddUser(context.Background(), userUUID, adminUser, userCreatorUuid)
+	c.Assert(err, jc.ErrorIsNil)
+
+	salt, err := auth.NewSalt()
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Set password hash.
+	err = st.SetPasswordHash(context.Background(), userUUID, "passwordHash", salt)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Check that the password hash was set correctly.
+	db := s.DB()
+
+	rowAuth := db.QueryRow(`
+SELECT disabled
+FROM user_authentication
+WHERE user_uuid = ?
+	`, userUUID)
+	c.Assert(rowAuth.Err(), jc.ErrorIsNil)
+
+	var disabled bool
+	err = rowAuth.Scan(&disabled)
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Check(disabled, gc.Equals, false)
+
+	row := db.QueryRow(`
+SELECT password_hash
+FROM user_password
+WHERE user_uuid = ?
+	`, userUUID)
+	c.Assert(row.Err(), jc.ErrorIsNil)
+
+	var passwordHash string
+	err = row.Scan(&passwordHash)
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Check(passwordHash, gc.Equals, "passwordHash")
+
+	row = db.QueryRow(`
```

Comment:
We should probably set an activation key for the user so we are checking the contract that it gets removed.

Reply:
agreed,

changed the code for adding the admin: 
```go
// Add admin user with activation key.
	userUUID, err := user.NewUUID()
	c.Assert(err, jc.ErrorIsNil)
	userCreatorUuid := jujuSystemUuid
	adminUser := user.User{
		Name:        "admin",
		DisplayName: "admin",
	}
	err = st.AddUserWithActivationKey(context.Background(), userUUID, adminUser, userCreatorUuid, "activationKey")
	c.Assert(err, jc.ErrorIsNil)
```